### PR TITLE
Added android instrumented test environment

### PR DESCRIPTION
**Pull request by aeharvlee**
## Proposed changes
This PR introduces **android instrumented test environment**.
We need to test and know whether `caver-java-andorid` lib 
is well-working on real android machine or not.

For this

**1. Added android_instrumented_test sub-project**
This sub-project `android_instrumented_test`  is created for instrumented testing.
This sub-project will use unit test files already written at core. 

This sub-project uses [AndroidX Test Library](https://developer.android.com/training/testing/unit-testing/instrumented-unit-tests) for instrumented test.

**2. Added Suite Test for Test classes containing nested classes**
Without it, Test classes containing nested classes does not work on emulator.

**3. Updated circle-ci config and added shell script**
To re-use unit tests written at core, I added a script for it and updated config of circle-ci for it.

**4. Updated build.gradle of root-project**
* Excludes `android_instrumented_test` from publish scope
* Excludes `android_instrumented_test` from scope applying java plugin

**5. Updated git ignore**
* Add [gitignore](https://github.com/github/gitignore/blob/master/Android.gitignore) of Android Project. 
* Remove `gradle.properties` from gitignore because it contains essential syntaxes for AndoirdX Test Library.

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/caver-java/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/caver-java)
- [x] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

- Please leave the issue numbers or links related to this PR here.

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: core/src/test/java/com/klaytn/caver/ipfs/IPFSTest.java [hunk 1]
```diff
@@ -81,6 +81,9 @@ public void toHex() {
 
     @Test
     public void addFile() throws IOException {
+        if (System.getenv("LOCAL_JVM") == null) {
```
Comment (sirano11) [inline core/src/test/java/com/klaytn/caver/ipfs/IPFSTest.java:84]: what does this code mean? 
When I wants to test, do I need to set the LOCAL_JVM environment variable? 
How was it intended to work in the Circle CI environment?

File: .gitignore [hunk 2]
```diff
@@ -26,7 +26,6 @@
 .idea/**/gradle.xml
 .idea/**/libraries
 .gradle/
-gradle.properties
```
Comment (sirano11) [inline .gitignore:29]: Why did you remove gradle.properties from .gitignore file?

File: android_instrumented_test/build.gradle [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+    compileOptions {
+        sourceCompatibility 1.8
+        targetCompatibility 1.8
+    }
+}
+
+dependencies {
+    implementation project(':core')
+    androidTestImplementation "androidx.test.ext:junit:1.1.2"
```
Comment (sirano11) [inline android_instrumented_test/build.gradle:42]: Please set a same format that presented library version with external variable.(Check root project's build.gradle)

File: .circleci/config.yml [hunk 4]
```diff
@@ -56,7 +62,16 @@ build_test: &build_test
...
+        ./gradlew build --debug -x :android_instrumented_test:build
+
+build_test_android: &build_test_android
+  run:
+    name: "Build and test"
+    no_output_timeout: 20m
+    command: |
+      ./gradlew clean
+      ./gradlew connectedAndroidTest
+      ./gradlew build --debug -x :android_instrumented_test:build -x test
```
Comment (sirano11) [inline .circleci/config.yml:74]: Why did you build a source after test?

File: core/src/test/java/com/klaytn/caver/ipfs/IPFSTest.java (hunk 1 above)
Comment (aeharvlee) [reply core/src/test/java/com/klaytn/caver/ipfs/IPFSTest.java:84]: @sirano11 
Hi sirano :)
I added new Environment Variable "LOCAL_JVM" to circle ci config.
This env-variable is used to specify the environment test is being executed.

If `LOCAL_JVM` is set, it means "JVM" environment.
Else it means "Android" environment.

```
jobs:
  build:
    <<: *machine_ubuntu
    steps:
      - checkout
      - *pull_klaytn_image
      - *start_test_network
      - *check_test_network
      - run:
          name: "Check rc version"
          command: |
            if [ -z "$CIRCLE_TAG" ]; then
              echo "this is not RC version"
            else
              sed -i "s/version '.*'/version '${CIRCLE_TAG:1}'/" build.gradle
              awk '/version '\''/' build.gradle
            fi
            export LOCAL_JVM=true // HERE, but this may not an appropriate position. I think I should move this statement outside of `Check rc version` block.
```

1. If a test is running on a machine which has "LOCAL_JVM" environment variable and it contains `if` statement about `LOCAL_JVM` and it's condition is true, it means "This test case can only be running on JVM, not an Android Emulator."
2. If a test is running on a machine which has no `if` statement about "LOCAL_JVM" environment variable, it means "This test case can be running on both JVM and Android Emulator"

So with this `LOCAL_JVM`, we can make an android emulator skip test cases which can be only ran in JVM.

File: .gitignore (hunk 2 above)
Comment (aeharvlee) [reply .gitignore:29]: To make AndroidX Test library works, we need to add some statements about it in `gradle.properties` and it must be included in the project.
You can check this out [here](https://developer.android.com/jetpack/androidx#using_androidx_libraries_in_your_project) :)

File: core/src/test/java/com/klaytn/caver/ipfs/IPFSTest.java (hunk 1 above)
Comment (sirano11) [reply core/src/test/java/com/klaytn/caver/ipfs/IPFSTest.java:84]: So, How can i test a unit test in local? 
I think this code is focused only circle ci system.

File: .circleci/config.yml [hunk 5]
```diff
@@ -13,14 +13,20 @@ machine_ubuntu: &machine_ubuntu
 android_machine_ubuntu: &android_machine_ubuntu
   working_directory: ~/circleci-caver-java-android
   machine:
-    image: ubuntu-1604:201903-01
+    image: android:202102-01
```
Comment (whoisxx) [inline .circleci/config.yml:16]: is this image support by circleci?

Comment (aeharvlee) [reply .circleci/config.yml:16]: @whoisxx 

Yeap. 
you can check it here "[CircleCI Android machine image preview docs](https://github.com/CircleCI-Public/android-image-preview-docs#no-orb-example)" :)

File: .circleci/config.yml [hunk 6]
```diff
@@ -75,7 +90,10 @@ jobs:
               sed -i "s/version '.*'/version '${CIRCLE_TAG:1}'/" build.gradle
               awk '/version '\''/' build.gradle
             fi
+            export LOCAL_JVM=true
```
Comment (whoisxx) [inline .circleci/config.yml:93]: I think it better to use circleci environment values instead of export command here

File: .circleci/config.yml (hunk 4 above)
Comment (aeharvlee) [reply .circleci/config.yml:73]: @sirano11 
I adjusted the order of steps as described in the name of the step :)

F.Y.I
Here is the description about gradle commands about `android_instrumented_test` sub-project.
```
Build tasks
-----------
android_instrumented_test:build - Assembles and tests this project.
android_instrumented_test:connectedDebugAndroidTest - Installs and runs the tests for debug on connected devices.
```

File: .gitignore (hunk 2 above)
Comment (aeharvlee) [reply .gitignore]: @sirano11 

I added again `gradle.properties` at this [commit](https://github.com/klaytn/caver-java/pull/240/commits/466248cff26478e2ebb20ad0c5c2c7fe7ae1b62f) 😃

File: .circleci/config.yml (hunk 6 above)
Comment (aeharvlee) [reply .circleci/config.yml]: @whoisxx 
I removed this statement from config at this [commit](https://github.com/klaytn/caver-java/pull/240/commits/466248cff26478e2ebb20ad0c5c2c7fe7ae1b62f).
So it's okay not to check this part.

Thanks for your feedback.

If you want to know the reason why I used this environment variable, please read the below.
-------------
What I tried is to declare environment variable in Android Emulator and skip some tests based on this, but couldn't find about declaring env variable in Android emulator when testing.

So I did reversely. Declare env variable at host machine(Ubuntu).
But if we accept this idea, we must set env variable `LOCAL_JVM` every time we test on JVM. That's quite cumbersome and difficult to remember.

**Review by sirano11 — DISMISSED** [review]

File: .circleci/config.yml (hunk 6 above)
Comment (whoisxx) [reply .circleci/config.yml]: Understand. thank you for explanation

**Review by whoisxx — DISMISSED** [review]

**Review by sirano11 — APPROVED** [review]

**Review by whoisxx — APPROVED** [review]
LGTM
